### PR TITLE
feat(di)!: add pseudo orphan rule for framework type override prevention

### LINKED ISSUE #3463: feat(di): add pseudo orphan rule to prevent framework type override in injectable factories
## Summary

Add a pseudo orphan rule that prevents users from registering `#[injectable_factory]` or `#[injectable]` for framework-managed types (any type defined in the `reinhardt` workspace). Uses `std::any::type_name` prefix matching at startup validation.

## Problem

When users register factories for framework-managed types like `DatabasePool` or `HttpRequest`:

1. **Semantic correctness** — Framework types have their own initialization lifecycle. User-defined factories may cause double initialization or lifecycle management conflicts.
2. **API design violation** — Analogous to Rust's orphan rule, users should not override the framework's internal type management.
3. **DI container consistency** — Framework types may already be registered internally, creating TypeId conflicts (related to #3458).

## Proposed Solution

- Add `is_framework_type()` function using `std::any::type_name` prefix matching (`reinhardt::` and `reinhardt_`)
- Add `ValidationErrorKind::FrameworkTypeOverride` to `RegistryValidator`
- Hard error (startup panic) with clear error message and newtype wrapper hint
- 53 test cases covering normal/abnormal/edge cases/boundary values

## Design Document

`docs/superpowers/specs/2026-04-11-di-framework-type-orphan-rule-design.md`

## Breaking Change

This is a **breaking change** for any existing code that registers factories for framework types. Users must wrap framework types in newtypes.

## Files to Modify

- `crates/reinhardt-di/src/validation.rs`
- `crates/reinhardt-di/src/registry.rs`

### PR DESCRIPTION
**Pull request by kent8192**
## Summary

- Add pseudo orphan rule that prevents users from registering `#[injectable_factory]` or `#[injectable]` for framework-managed types
- Detection uses `std::any::type_name` prefix matching (`reinhardt::` / `reinhardt_`) at startup validation
- Violations cause a startup panic with clear error message and newtype wrapper migration hint
- 53 test cases covering normal/abnormal/edge cases and `type_name` format regression

## Breaking Change

This is a **breaking change** approved via SP-3 ([API Change Proposal #3465](https://github.com/kent8192/reinhardt-web/issues/3465)).

Any existing code that registers factories for framework types (e.g., `DatabasePool`, `HttpRequest`) will panic at startup. Users must wrap framework types in newtypes:

```rust
// Before (now panics):
#[injectable_factory]
async fn create_pool() -> DatabasePool { ... }

// After:
pub struct AppDatabasePool(pub DatabasePool);

#[injectable_factory]
async fn create_pool() -> AppDatabasePool { ... }
```

## Implementation Details

| Component | Change |
|-----------|--------|
| `DependencyRegistry` | New `qualified_type_names` map + accessor methods |
| `RegistryValidator` | New `FrameworkTypeOverride` variant + `check_framework_type_override()` |
| `#[injectable_factory]` macro | Calls `register_qualified_type_name()` with `std::any::type_name::<T>()` |
| `#[injectable]` macro | Same as above |
| `format_validation_report()` | Handles new error kind |

## Test plan

- [x] 12 abnormal cases (framework types correctly detected)
- [x] 13 normal cases (user/std/third-party types allowed)
- [x] 15 edge cases (similar prefixes, generics, case sensitivity, empty strings)
- [x] 8 validation integration tests (single/multiple violations, error messages)
- [x] 3 `type_name` format regression tests
- [x] All 145 existing reinhardt-di tests pass
- [x] Clippy clean, rustfmt clean

Fixes #3463

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: crates/reinhardt-di/src/validation.rs [hunk 1]
```diff
@@ -183,6 +198,29 @@ impl RegistryValidator {
...
+				let display_name = self.resolve_type_name(*type_id);
+				errors.push(ValidationError {
+					kind: ValidationErrorKind::FrameworkTypeOverride,
+					type_name: display_name,
+					type_id: *type_id,
+					message: format!(
+						"Type `{qualified_name}` is a framework-managed type and cannot be \
+						 registered via #[injectable_factory] or #[injectable]. \
+						 Framework-managed types are automatically provided by the framework. \
+						 Help: Define your own wrapper type instead."
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-di/src/validation.rs]: Valid point. The `\` line continuation with indentation was embedding unwanted spaces in the rendered message. Fixed by switching to `concat!()` with positional `{}` placeholder (since `concat!` expands before `format!`, named captures like `{qualified_name}` don't work through `concat!`).

---
🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: crates/reinhardt-di/src/registry.rs [hunk 2]
```diff
@@ -271,6 +275,27 @@ impl DependencyRegistry {
...
+	pub fn get_qualified_type_name(&self, type_id: &TypeId) -> Option<&'static str> {
+		self.qualified_type_names.get(type_id).map(|r| *r.value())
+	}
+
+	/// Get all qualified type name mappings.
+	pub fn get_all_qualified_type_names(&self) -> std::collections::HashMap<TypeId, &'static str> {
+		self.qualified_type_names
+			.iter()
+			.map(|entry| (*entry.key(), *entry.value()))
+			.collect()
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-di/src/registry.rs]: Agreed. Replaced `get_all_qualified_type_names()` (which cloned the entire `DashMap` into a `HashMap`) with `iter_qualified_type_names()` returning `impl Iterator<Item = (TypeId, &'static str)> + '_`. The validation loop now iterates directly over the `DashMap` entries without allocation.

---
🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: crates/reinhardt-di/src/registry.rs [hunk 3]
```diff
@@ -271,6 +275,27 @@ impl DependencyRegistry {
 	pub(crate) fn get_type_name(&self, type_id: TypeId) -> Option<&'static str> {
 		self.type_names.get(&type_id).map(|entry| *entry.value())
 	}
+
+	/// Register the fully-qualified type name obtained from `std::any::type_name::<T>()`.
+	///
+	/// Used by the pseudo orphan rule to detect framework-managed types.
+	#[doc(hidden)]
+	pub fn register_qualified_type_name(&self, type_id: TypeId, qualified_name: &'static str) {
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-di/src/registry.rs:310]: This method must remain `pub` (not `pub(crate)`). The `#[injectable]` and `#[injectable_factory]` proc macros generate code that calls `register_qualified_type_name` from user crate context (see `crates/reinhardt-di/macros/src/injectable.rs:338` and `crates/reinhardt-di/macros/src/injectable_factory.rs:186`). Since proc macros expand in the downstream crate, `pub(crate)` visibility would cause compilation errors for all macro users.

The `#[doc(hidden)]` attribute already signals this is not part of the public API. The orphan rule validation itself catches any misuse at startup, so the enforcement concern is addressed.

---
🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: crates/reinhardt-di/src/validation.rs [hunk 4]
```diff
@@ -40,11 +42,23 @@ impl fmt::Display for ValidationError {
...
 
+/// Check if a type belongs to the reinhardt framework based on its
+/// fully-qualified name from `std::any::type_name`.
+///
+/// Returns `true` for types whose qualified name starts with `reinhardt::`
+/// (the facade crate) or `reinhardt_` (any sub-crate like `reinhardt_di`,
+/// `reinhardt_db`, etc.).
+fn is_framework_type(qualified_name: &str) -> bool {
+	qualified_name.starts_with("reinhardt::")
+		|| (qualified_name.starts_with("reinhardt_") && qualified_name.len() > "reinhardt_".len())
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-di/src/validation.rs]: Valid concern. Replaced the broad `reinhardt_` prefix matching with an explicit `FRAMEWORK_CRATE_PREFIXES` allowlist constant containing all known framework crate prefixes (e.g., `"reinhardt_di::"`, `"reinhardt_db::"`, etc.). This eliminates false positives for third-party crates that happen to start with `reinhardt_`.

The allowlist is defined as a `const &[&str]` at module level with a doc comment instructing maintainers to update it when adding new framework crates. Updated the edge-case test accordingly: `reinhardt_like_crate::MyType` now correctly returns `false`.

---
🤖 Generated with [Claude Code](https://claude.com/claude-code)
